LLVM panic at compilation when parapoly type is incorrect

## Context

* Please paste `odin report` output:
```
        Odin:    dev-2024-08-nightly:8359995
        OS:      macOS Sonoma 14.3.1 (build: 23D2057, kernel: 23.3.0)
        CPU:     Apple M3
        RAM:     16384 MiB
        Backend: LLVM 18.1.8
```

## Expected Behavior

Compiler will print normal error message.

## Current Behavior

I get LLVM panic

## Failure Information (for bugs)

### Steps to Reproduce

Run this:

```odin
package main

Container :: struct($T: typeid) {
	item: T,
}

new_container :: proc() -> Container($T) {
	return Container(T){}
}

main :: proc() {
	c := new_container()
}
```

### Failure Logs

```
src/llvm_backend_proc.cpp(72): Panic: new_container :: proc() -> Container($T) (was parapoly: 1 1)
make: *** [run] Trace/BPT trap: 5
```

--- a/src/check_type.cpp
+++ b/src/check_type.cpp
@@ -2440,8 +2440,12 @@ gb_internal bool check_procedure_type(CheckerContext *ctx, Type *type, Ast *proc
 	bool success = true;
 	isize specialization_count = 0;
 	Type *params  = check_get_params(c, c->scope, pt->params, &variadic, &variadic_index, &success, &specialization_count, operands);
-	Type *results = check_get_results(c, c->scope, pt->results);
 
+	bool no_poly_return = c->disallow_polymorphic_return_types;
+	c->disallow_polymorphic_return_types = c->scope == c->polymorphic_scope;
+	// NOTE(zen3ger): if the parapoly scope is the current proc's scope, then the return types shall not declare new poly vars
+	Type *results = check_get_results(c, c->scope, pt->results);
+	c->disallow_polymorphic_return_types = no_poly_return;
 
 	isize param_count = 0;
 	isize result_count = 0;
@@ -3383,6 +3387,9 @@ gb_internal bool check_type_internal(CheckerContext *ctx, Ast *e, Type **type, T
 		}
 		Type *t = alloc_type_generic(ctx->scope, 0, token.string, specific);
 		if (ctx->allow_polymorphic_types) {
+			if (ctx->disallow_polymorphic_return_types) {
+				error(ident, "Undeclared polymorphic parameter '%.*s' in return type", LIT(token.string));
+			}
 			Scope *ps = ctx->polymorphic_scope;
 			Scope *s = ctx->scope;
 			Scope *entity_scope = s;

--- a/src/checker.hpp
+++ b/src/checker.hpp
@@ -521,6 +521,7 @@ struct CheckerContext {
 	bool       in_enum_type;
 	bool       collect_delayed_decls;
 	bool       allow_polymorphic_types;
+	bool       disallow_polymorphic_return_types; // NOTE(zen3ger): no poly type decl in return types
 	bool       no_polymorphic_errors;
 	bool       hide_polymorphic_errors;
 	bool       in_polymorphic_specialization;

